### DISCUSSION TITLE
Improved/type-safe PSBT API

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#1440] [Ideas]
As we were already discussing the current dynamic nature of PSBTs makes them complicated and it makes it hard to maintain all required invariants. I suggested that every consumer should define their own (new)types depending on their needs. And while I still believe that's the case maybe there's a way to help.

My idea consists of three parts:

* Implement the existing proposal to split `unsigned_tx` into individual inputs, outputs etc.
* Provide additional primitives such as `enum PrevOut { SegWit(TxOut), Legacy(Transaction) }`, these are not to be used manually
* Provide a proc macro (IMO too complicated and ugly for `macro_rules`) that can verify certain invariants such as presence of required fields.

Rough example:

```rust
// We're defining input for our custom PSBT
// `allow_extra_fields` will cause the conversion from raw PSBT to not fail if an unknown field is encountered
#[psbt(Input, allow_extra_fields)]
struct MyPsbtInput {
    // The proc macro recognizes certain names of the fields, unknown fields are rejected
    // The proc macro also recognizes specific types (which don't have to be imported), unknown types are rejected
    prev_out: PrevOut, // You could force legacy by using `Transaction` type. If prev out is allowed but optional
    // imagined type, not sure if it should look like this
    script_sig: Option<ScriptBuf>,
}

#[psbt(Psbt)]
struct MyPsbt {
    inputs: Vec<MyPsbtInput>,
    // Definition omitted for clarity
    outputs: Vec<MyPsbtOutput>,
}
```

The macro transforms the structs a bit and provides conversion to/from generic PSBT or direct (de)serialization.
Applications that work with multiple forms of PSBT (e.g. one fully signed and one partially signed in BIP78) can define multiple types.
This bridges the dynamic nature of PSBTs with static type checks provided by Rust.

What do you think? Disclaimer: my thinking is not great lately so maybe this is a bad idea. :)

Cc @DanGould and @dr-orlovsky who seem to be interested in the topic of PSBTs.

**Comment by tcharding** [comment]
What led you to thinking that users defining the types is required? Said another way, are there not a set of types that can be used as all building blocks for PSBTs?

**Reply by Kixunil** [reply]
For instance BIP78 mandates that certain fields are **not** present in PSBT. So if you use our dynamic PSBT you have to manually check that. It also (implicitly) requires that prevouts are known if I'm not mistaken. In `bip78` crate there's a newtype around `Psbt` marking that required checks were performed but internally it still contains the dynamic PSBT. So "make invalid state unrepresentable" isn't perfectly implemented. It'd be great to do it perfectly but that requires a bunch of boilerplate which my idea should reduce.

And I have a very similar issue in some other WIP smart contract.

**Comment by dr-orlovsky** [comment]
Thank you for tagging. I diverged from rust-bitcoin PSBT implementation some time ago and maintain own PSBT crate by LNP/BP Association: https://crates.io/crates/psbt, the source is in https://github.com/BP-WG/descriptor-wallet/tree/master/psbt/src

Specifically, it separates unsigned_tx into input/output fields, allows arbitrary proprietary keys and other functionality.

**Reply by Kixunil** [reply]
Yeah, I thought you may find the idea of creating domain-specific ad-hoc PSBT types using proc macros interesting.

**Reply by dr-orlovsky** [reply]
Generally I do not like proc macros where they can be avoided. Maybe better to have a method `Psbt::validate` which takes a `impl PsbtValidator` custom type and not to derive anything?

**Reply by Kixunil** [reply]
Me neither. The reason for proc macro is that you can do validation today by implementing `TryFrom<Psbt> for Newtype` but then you either have to wrap the dynamic `Psbt` and litter the code with `unwrap`s or manually write your own struct holding the required fields.

**Reply by dr-orlovsky** [reply]
So why not to use a dedicated verifier type with a well-defined API?

```rust
pub trait Verifier {
    type Error: std::error::Error;
    type Newtype: Deref<Psbt>;
    fn verify_input(&self, input: &Input) -> Result<(), Self::Error>;
    fn verify_output(&self, output: &Output) -> Result<(), Self::Error>;
}

impl Psbt {
   // ...
   pub fn verify<V: Verifier>(&self, verifier: &V) -> Result<V::Newtype, Vec<V::Error>> {
     // run verifier over each input and output
   }
}
```

**Reply by Kixunil** [reply]
Let's say my application requires `seuence_number` to be set for all inputs and the verification needs to be done before processing the PSBT. If I do it the way you suggested I still have to put `sequence_numer.unwrap()` into the business logic which is fragile. I'd rather redefine the `Input` to contain the field without `Option`.

**Reply by dr-orlovsky** [reply]
Ok, I see. What can I say from my experience about application-specific PSBTs: usually a binary does several "applications" requiring different fields in PSBT to be present. I assume having multiple types will end up with too much code generated by macros.

My approach would be to have a field-specific methods returning `Result`s, and do `psbt.sequence_number()?`, to do a lazy validation when required.

PSBT standards defines that PSBT with `sequence_number` not set is a valid one, so I think that application-specific type safety here should be replaced with runtime validation using result monads.

**Reply by Kixunil** [reply]
> too much code generated by macros.

I have an experience with the other extreme (but it's a complicated toml, not PSBT) and the result is absolutely horrible. Bunch of runtime validation checks in multiple places that are inconsistent, hard to to undertand, hard to refactor, hard to extend (to support multiple errors and spans) and are in fact duplication of code or panics (panics are acceptable in the projects but the desire is still to change them to proper spanned reports long-term). I've wished hundreds of times to go back in time, slap myself into face and say "do it properly, dumbass!".

Would **not** recommend. I'd rather write the additional structs and verification code by hand than ever repeat this again.

`sequence_number` is also just a silly example. If there's one trivial check like that maybe it's fine to check inside business rules but for more complex stuff I'd stay away. *Especially* in security-critical software.

**Comment by DanGould** [comment]
At least the first two points of this proposal are investments into the outdated v0 spec, since `unsigned_tx` is removed from v2 and the input / output map elements have all of the required data to extract a valid tx. Both the BIP78 issue and the sequence number issue you're describing are fixed by PSBTv2. 

I think the reason BIP78 psbt has to exclude unfinalized inputs' partial signature fields is because of the clunky v0 input finalizer role. In v2 it does not matter if inputs are finalized because there is no finalizer role. A bip78 protocol psbtv2 upgrade that might work could be for the sender to check for the absence of PSBT_GLOBAL_TX_MODIFIABLE=True and then proceed with the signer which has sufficient validation logic. The contents of fields need to be checked before signing, so I'm not sure the proc_macro, which only checks the shape of the data iiuc, helps enough to merit the complexity.

I think the PSBT_GLOBAL_TX_MODIFIABLE=True absence could be set/checked for your second sequence example too. Just set the field to false or remove it once sequence is set for all inputs.

I admit proc macros are a blind spot for me, so I don't understand the whole of your idea. Specifically, what problem does it solve? It smells roughly like a want for typestate to ready for each PSBT role.

**Reply by Kixunil** [reply]
> Both the BIP78 issue and the sequence number issue you're describing are fixed by PSBTv2.

How is PBTv2 affecting Rust type system? Unless it makes all fields mandatory which doesn't make sense.

> It smells roughly like a want for typestate to ready for each PSBT role.

Pretty much yes and not just role but also for custom applications that e.g. interactively combine multiple inputs/outputs.

Maybe to explain it better, consider (PSBT-unrelated) [code from `configure_me_codegen`](https://github.com/Kixunil/configure_me/blob/be885a53d79a74a5e961f847a05c54bee5fbeacb/configure_me_codegen/src/config.rs)

It's task is to parse and validate the fields in a toml file. Semantics of some of the fields is that they are contradictory though. Toml has no concept of enums, so there are separate structs that do "dumb" deserialization which maps 1:1 with the underlying format. E.g. `raw::Param`

https://github.com/Kixunil/configure_me/blob/be885a53d79a74a5e961f847a05c54bee5fbeacb/configure_me_codegen/src/config.rs#L679-L697

which maps to `Param`:

https://github.com/Kixunil/configure_me/blob/be885a53d79a74a5e961f847a05c54bee5fbeacb/configure_me_codegen/src/config.rs#L896-L910

As you can see it has fewer fields and uses newtype `Ident` instead of `Spanned<String>` (technically `Ident` could impl `Deserialize` but that led to worse error messages when it did). The resulting types definitely valid contain configuration specification.

So it's a kind of typestate with two types where state transition is validation. But now I realized the problem is more general than PSBTs since it's also present in `configure_me` (BTW it's there twice - same thing in the generated code) and I think I encountered it somewhere else too. And also I remember I was considering creating some proc macro validator but concluded the rules can be so complex that the amount of code you'd write with pro macros is roughly the same as manually anyway.

So I think I will just not use PSBTs in the protocol I'm designing. I may use them for communication with other applications though.

**Reply by DanGould** [reply]
> How is PBTv2 affecting Rust type system? Unless it makes all fields mandatory which doesn't make sense.

Only v0 has the terrible invariant problem. Because of the fixed `unsigned_tx` role order is very strict, and might benefit from added type  complexity. v2 has no invariant `unsigned_tx`, so validation can be handled by the signer for certain after 1 flag is set.

**Reply by Kixunil** [reply]
I don't understand, aren't some fields still optional?

**Reply by DanGould** [reply]
Yes, but the validation in v2 is much less of an issue in the way it is for PayJoin because of v0. The way I see it, having nested data in two separate and a busload of roles causes most of the problems. PSBTv2's simple Constructor / signer / extractor roles fix 80% of the problem.

I wrote a short piece to address your concern: https://chaincase.app/words/interactive-transactions-psbt

**Reply by Kixunil** [reply]
Indeed, it's much less of an issue but I think it still can be improved by tooling. It may be the case that the ROI is not worth it at this point though.

Anyway, in my project I dropped use of PSBT for this reason - I control both sides so I can just use fixed encoding.

**Reply by DanGould** [reply]
what did you replace it with?

**Reply by Kixunil** [reply]
Manual struct serialization as bytes. It is very similar to consensus encoding. I encode Bitcoin types using consensus encoding and everything else as (big endian) bytes. For vecs I encode 32-bit big endian length followed by items. Note that many fields I transmit aren't even part of PSBT spec. I could use custom keys but still there are messages that contain multiple transactions.

Also I don't transmit everything, just the required information - each participant re-assembles the transactions deterministically from received data. This way only minimal verification is required (since other data is already trusted or verified) and there's great confidence there won't be weird edge similar to some past bugs in LN.

This whole thing will need a bit more code to support versioning and extensions but shouldn't be too hard.
